stablehlo/reference/Ops.cpp

```diff
@@ -102,6 +102,123 @@ Tensor evalClampOp(const Tensor &min, const Tensor &operand, const Tensor &max,
   return result;
 }
 
+Tensor evalCompareOp(const Tensor &lhs, const Tensor &rhs,
+                     ComparisonDirection comparisonDirection,
+                     std::optional<ComparisonType> compareType,
+                     TensorType resultType) {
+  Tensor result(resultType);
+  auto elementTy = result.getType().dyn_cast<TensorType>().getElementType();
+  auto comparisonType = ComparisonType::NOTYPE;
+  if (compareType && *compareType != ComparisonType::NOTYPE) {
+    comparisonType = *compareType;
+  } else if (isSupportedUnsignedIntegerType(elementTy) ||
+             isSupportedBooleanType(elementTy)) {
+    comparisonType = ComparisonType::UNSIGNED;
+  } else if (isSupportedSignedIntegerType(elementTy)) {
+    comparisonType = ComparisonType::SIGNED;
+  } else if (isSupportedFloatType(elementTy) ||
+             isSupportedComplexType(elementTy)) {
+    comparisonType = ComparisonType::FLOAT;
```

Comment:
Hmm, my opinion is that the comment is not necessary. There are 5 enum types, namely `NOTYPE`, `UNSIGNED`, `SIGNED`, `FLOAT`, and `TOTALORDER`. Obviously, `NOTYPE`, `SIGNED`, and `UNSIGNED` are out of the question. It can't be `TOTALORDER` since this has to do with the absolute ordering of the elements and unrelated to the element type of what is being compared. Complex types shouldn't be treated differently, and under the hood, it uses a pair of floating-points. The spec also does state it in (C3):
```
* (C3) Given `E` is the `lhs` element type, the following are legal values of
       `compare_type`:
  * ...
  * ...
  * ...
  * If `E` is complex type, `compare_type` = `FLOAT`.
```

Reply:
Agreed that the spec is clear about what comparison type should be used for complex types. It is still not obvious to me why we should be using `FLOAT` over `TOTALORDER` for floating-point types, since the spec says that `TOTALORDER` is (only) legal for them.

Also, now that I'm thinking further about this, the spec doesn't say anything about NOTYPE, so we need a comment anyway because even computing a default value of ComparisonType is deviating from the spec. How about this?

```C++
  // TODO(#1311): Unlike `stablehlo.compare` from the spec, the StableHLO 
  // dialect defines `compare_type` as an optional attribute, so here we 
  // explicitly compute its default value.
  // For non-floating point element types, the default value is unambiguously
  // defined from the C3 constraint in the spec. For floating-point element
  // types, we're picking FLOAT because that's the conventional semantics
  // across the users of CompareOp.
  auto comparisonType = ComparisonType::NOTYPE;
  if (compareType) {
    ...
```

Reply:
I see your point, your suggestion makes sense.